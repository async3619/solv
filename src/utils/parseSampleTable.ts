import { HTMLElement, Node, NodeType } from "node-html-parser";
import { decode } from "html-entities";

import { Challenge } from "./types";

const chunk = require("lodash.chunk");

function getAllChildren(element: Node, tagName: string, depth = 0) {
    const result: Node[] = [element];
    for (const node of element.childNodes) {
        result.push(...getAllChildren(node, tagName, depth + 1));
    }

    return depth === 0
        ? result
              .filter(e => element !== e)
              .filter(e => e.nodeType === NodeType.ELEMENT_NODE)
              .filter(e => (e as HTMLElement).tagName === tagName)
        : result;
}

export function parseSampleTable(table: HTMLElement): Pick<Challenge, "output" | "input"> {
    const head = table.querySelector("thead");
    const body = table.querySelector("tbody");
    if (!head || !body) {
        throw new Error("Given element is not a valid sample table object.");
    }

    const result: Pick<Challenge, "output" | "input"> = {
        input: [],
        output: [],
    };

    const columns = getAllChildren(head, "TH").map(node => node.textContent.trim());
    const rows = chunk(
        getAllChildren(body, "TD").map(node => node.textContent.trim()),
        columns.length,
    );

    for (let i = 0; i < rows.length; i++) {
        const row = rows[i];
        const inputValues: string[] = [];
        for (let j = 0; j < columns.length; j++) {
            if (j === columns.length - 1) {
                result.output.push(decode(row[j]));
                continue;
            }

            inputValues.push(row[j]);
        }

        result.input.push(inputValues);
    }

    return result;
}
