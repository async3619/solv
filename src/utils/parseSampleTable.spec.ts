import { parse as parseDOM } from "node-html-parser";

import { parseSampleTable } from "./parseSampleTable";

describe("parseSampleTable", () => {
    it("should parse sample table successfully", () => {
        const document = parseDOM(`<html>
  <body>
    <table class="table">
      <thead>
        <tr>
          <th>rectangles</th>
          <th>result</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>[[0, 1, 4, 4], [3, 1, 5, 3]]</td>
          <td>14</td>
        </tr>
        <tr>
          <td>
            [[1, 1, 6, 5], [2, 0, 4, 2], [2, 4, 5, 7], [4, 3, 8, 6], [7, 5, 9, 7]]
          </td>
          <td>38</td>
        </tr>
      </tbody>
    </table>
  </body>
</html>`);

        const table = document.querySelector("table");

        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        expect(parseSampleTable(table!)).toMatchSnapshot();
    });

    it("should throws an error when received corrupted table", () => {
        const document = parseDOM(`<html><body><table></table></body></html>`);
        const table = document.querySelector("table");

        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        expect(() => parseSampleTable(table!)).toThrow();
    });
});
