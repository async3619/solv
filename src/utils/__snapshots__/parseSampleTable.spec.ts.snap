// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseSampleTable should parse sample table successfully 1`] = `
{
  "input": [
    [
      "[[0, 1, 4, 4], [3, 1, 5, 3]]",
    ],
    [
      "[[1, 1, 6, 5], [2, 0, 4, 2], [2, 4, 5, 7], [4, 3, 8, 6], [7, 5, 9, 7]]",
    ],
  ],
  "output": [
    "14",
    "38",
  ],
}
`;
